RFTrain/src/model_factory.py

```diff
@@ -1,50 +1,92 @@
 
 
 class ModelFactory:
-    def __init__(self, model_config_path, model_type):
-        self.model_type = model_type
-        self.model_args = import_model_args(model_config_path, self.model_type)
+    def __init__(self, model_config_path, classifier_type, cv_folds=5, scoring_func='roc_auc', random_state=0):
+        self.classifier_type = classifier_type.lower()
+        self.gridsearch_args = import_gridsearch_args(model_config_path, self.classifier_type)
+        self.cv_folds = cv_folds
+        self.sampler_name = import_sampler(model_config_path)
+        self.preprocessor_name = import_preprocessor(model_config_path)
+        self.random_state = random_state
+        self.scoring_func = scoring_func
 
     def make_model(self):
-        type_to_maker = {
-            'RF': self.make_rf,
-            'SVM': self.make_svc,
-            'LSVM': self.make_lsvc,
-            'SGD': self.make_sgd,
-            'DNN': self.make_dnn,
-            'OCC': self.make_occ,
-            'GNB': self.make_gnb,
-            'CNB': self.make_cnb,
-        }
+        preprocessor = self.__make_preprocessor()
+        sampler = self.__make_sampler()
+        classifier = self.__make_classifier()
+        stages = []
 
-        return type_to_maker[self.model_type]()
+        if preprocessor is not None:
+            stages.append(('preprocessor', preprocessor))
+        if sampler is not None:
+            stages.append(('sampler', sampler))
 
-    def make_rf(self):
-        return RandomForestClassifier(**self.model_args)
+        stages.append(('classifier', classifier))
 
-    def make_svc(self):
-        return SVC(**self.model_args)
+        pipe = ImbPipeline(stages)
 
-    def make_lsvc(self):
-        return LinearSVC(**self.model_args)
+        stratified_kfold = StratifiedKFold(n_splits=self.cv_folds,
+                                           shuffle=True,
+                                           random_state=self.random_state)
 
-    def make_sgd(self):
-        return SGDClassifier(**self.model_args)
+        return GridSearchCV(estimator=pipe,
+                            param_grid=self.gridsearch_args,
+                            scoring=self.scoring_func,
+                            cv=stratified_kfold,
+                            n_jobs=-1,
+                            verbose=3)
 
-    def make_dnn(self):
-        pass  # TODO: Add implementation of custom keras-based neural network
+    def __make_preprocessor(self):
+        name_to_implementation = {
+            'MaxAbsScaler': preprocessing.MaxAbsScaler,
+            'MinMaxScaler': preprocessing.MinMaxScaler,
+            'Normalizer': preprocessing.Normalizer,
+            'RobustScaler': preprocessing.RobustScaler,
+            'StandardScaler': preprocessing.StandardScaler,
+        }
+        implementation = name_to_implementation.get(self.preprocessor_name, None)
+        if implementation is None:
```

Comment:
you can simplify it and return just `implementation()`
correct me if i'm wrong

Reply:
I could remake function so that `implementation()` would be a correct return. but currently it will raise errors